### PR TITLE
 feat(cmdb): methods to update a CMDB 

### LINKED ISSUE #30: <feature> Methods to update a CMDB
## Expected Behavior
Additional methods should be available on the CMDB File System to 
- create a directory (equivalent of ```mkdir``` with ```-p``` as an option)
- copy files between directories (equivalent of ```cp``` with ```-rp``` as an option)
- remove files and directories (equivalent of ```rm``` with ```-rf``` as options)
- create/append/overwrite a file ( equivalent of output redirection ```>``` and ```>>```)

Any cache information will optionally be synchronised with these changes.

The file path will use the CMDB File System file path, and any new file/directory will be created
- in the CMDB that matches the longest portion of the desired file path, and if more than one candidate ...
- the highest CMDB in the CMDB hierarchy.

## Current Behavior
The ```getCMDBTree()``` function currently only offers read access to CMDB files and directories. This change will round out its file support.

## Possible Solution
The extra functionality could be covered by four new CMDB related functions as shown below. Where possible, the corresponding unix command behaviour should define the behaviour of these methods. This should permit a straightforward conversion of these requests to underlying operating system rquests.

**mkdirCMDB** - create directory

```
[#local result =
  mkdirCMDB(
    path,
    {
      "Parents" : false,
      "Synch" : true
    }
  ) ]
```
where
- path is the path within the CMDB file system for the directory to be created
- the last parameter is a hash of options
    - Parents is the equivalent of ```-p``` and will create any missing directories in the path (default false).
    - Synch will cause any cached information to be synchronised (default true)
- the function returns the status of the operation consistent with unix status codes ( 0 = success).

**cpCMDB** - copy files/directories

```
[#local result =
  cpCMDB(
    from,
    to,
    {
      "Recurse" : false,
      "Preserve" : false,
      "Synch" : true
    }
  ) ]
```
where
- from is the glob within the CMDB File System representing the files/directories that should be copied
- to is the target within the CMDB File System representing where the files or directories should be copied
- the last parameter is a hash of options
    - Recurse is the equivalent of ```-r``` and will recursively copy directories as well as files (default false).
    - Preserve is the equivalent of ```-p``` and will preserve the file attributes of existing files (default false).
    - Synch will cause any cached information to be synchronised  (default true).
- the function returns the status of the operation consistent with unix status codes ( 0 = success).

Note that this method will need to support copying of binary and text files.

**rmCMDB** - remove files/directories

```
[#local result =
  rmCMDB(
    glob,
    {
      "Recurse" : false,
      "Force" : false,
      "Synch" : true
    }
  ) ]
```
where
- glob is the glob within the CMDB file system representing the files or directories to be deleted
- the last parameter is a hash of options
    - Recurse is the equivalent of ```-r``` and will recursively delete directories as well as files (default false).
    - Force is the equivalent of ```-f``` and will not error on non-existent files or directories (default false).
    - Synch will cause any cached information to be synchronised (default true).
- the function returns the status of the operation consistent with unix status codes ( 0 = success).

**toCMDB** - write a file

```
[#local result =
  toCMDB(
    path,
    content,
    {
      "Append" : false,
      "Synch" : true,
      "Format" : "json"
    }
  ) ]
```
where
- path is the path within the CMDB file system to the file to be created/updated
- content is a freemarker object to be written to the file. The provided content should thus be serialised per the Format option, but arbitrary textual content can be written via use of a string as the content. 
- the last parameter is a hash of options
    - Append is the equivalent of ```>>``` and will cause the content to be appended to any existing file.  (default false = overwrite).
    - Synch will cause any cached information to be synchronised (default true).
    - Format is one of "json", "yaml" or "yml". The content will be serialised in the format requested (default json) (see also https://github.com/hamlet-io/engine-core/issues/32)
- the function returns the status of the operation consistent with unix status codes ( 0 = success).

**fromCMDB** - read directories/files

For consistency, ```getCMDBTree()``` could be renamed to ```fromCMDB```, or an additional method added. Note that the return value would continue to be the results of the read, and NOT the status of the operation.

## Context
This issue fleshes out https://github.com/hamlet-io/engine/issues/1265, and is initially intended to support migration of the cmdb upgrade/cleanup process to be in the engine rather than in the executor.

It also opens the possibility of generating multiple outputs in a single invocation of the engine.

## Your Environment

* CMDB Version used: All

### PR DESCRIPTION
**Pull request by kshychko**
## Motivation and Context

Fixes #30
## How Has This Been Tested?

On local environment
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions

- [x] None

## Checklist:

- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] None of the above.

### CODE REVIEW

**Comment by ml019** [comment]
@kshychko let's not check in all the rc versions to minimise the repo size. They are also redundant with the publication of 1.12,

**Comment by roleyfoley** [comment]
> @kshychko let's not check in all the rc versions to minimise the repo size. They are also redundant with the publication of 1.12, 

Could we maybe setup releases with a GitHub action to build these?

**Comment by ml019** [comment]
@kshychko I thought we incorporated the write support with the logging control change. Is this PR still valid or is all the update support in place but needing more testing?

**Comment by ml019** [comment]
@roleyfoley would appreciate your advice as to how to set up releases via github actions. I'm assuming we'd use a tag convention to drive this?

**Comment by roleyfoley** [comment]
> @roleyfoley would appreciate your advice as to how to set up releases via github actions. I'm assuming we'd use a tag convention to drive this?

Ticket raised https://github.com/hamlet-io/engine-core/issues/39 we should cover it separately from this PR

**Comment by kshychko** [comment]
@ml019, the core is using 1.12 version, when this PR includes the fix for mkdir 996e26e6f9346437865af80d234c2553d1244043 that comes iwht 1.12.1-rc1 version and I don't think this one was tested.

**Comment by kshychko** [comment]
@ml019, I rebased this PR to include changes from #40 so you can `freemarker-wrapper-1.12.1-rc2.jar` (packed as artifact `freemarker-wrapper.jar.zip`) from `Java CI with Maven` check for the latest commit.

**Comment by kshychko** [comment]
and I added one more commit 531e251, bumped version to 1.12.1-rc3

**Comment by ml019** [comment]
@kshychko I don't see the build in the releases section with the tag but maybe I'm looking in the wrong place? Basically where do I go to get the version corresponding to this PR?

**Comment by roleyfoley** [comment]
> @kshychko I don't see the build in the releases section with the tag but maybe I'm looking in the wrong place? Basically where do I go to get the version corresponding to this PR?

Looks like the latest build is available under the draft release 
https://github.com/hamlet-io/engine-core/releases 

There is a draft release at the top, expand the asssets and it should be that one.

**Comment by kshychko** [comment]
> @kshychko I don't see the build in the releases section with the tag but maybe I'm looking in the wrong place? Basically where do I go to get the version corresponding to this PR?

the current version from the PR is available as an artifact from the corresponding GitHub action run. So just click on a green tick next to the latest commit 531e251 or under `All checks have passed` and `Details` of ` Java CI with Maven / build (pull_request) `. This will get you to the page https://github.com/hamlet-io/engine-core/runs/1282473980 where the jar can be downloaded as an artifact.

**Comment by RossMurr4y** [comment]
@kshychko @ml019 It isn't clear if this PR is growing in scope or progressing towards the original ticket. Should this be marked as draft if it isn't ready for implementation?

**Comment by kshychko** [comment]
@RossMurr4y, the testing is in progress, but I agree that is better to have this PR as Draft.

File: freemarker-wrapper/pom.xml [hunk 1]
```diff
@@ -5,8 +5,8 @@
     <modelVersion>4.0.0</modelVersion>
 
     <artifactId>freemarker-wrapper</artifactId>
-    <groupId>io.codeontap</groupId>
-    <version>1.11.1</version>
+    <groupId>io.hamlet</groupId>
+    <version>1.12.1-rc10</version>
```
Comment (ml019) [inline freemarker-wrapper/pom.xml:9]: Does this need updating prior to formal release?

**Review by ml019 — COMMENTED** [review]
Just a minor query

File: freemarker-wrapper/pom.xml (hunk 1 above)
Comment (kshychko) [reply freemarker-wrapper/pom.xml]: @ml019, I update the version in the pom.xml file.

**Review by ml019 — APPROVED** [review]
Ready for release

**Pull request merged**